shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt

```diff
@@ -99,4 +101,143 @@ data class PatternsByStop(
+                    .mapValues { routeEntry ->
+                        routeEntry.value.groupBy({ it.first }, { it.second })
+                    }
+
+            val predictedPatternsByRoute =
+                staticData.patterns
+                    .filter {
+                        tripsByPattern[it.id]?.any { trip ->
+                            trip.prediction != null && trip.vehicle?.tripId == trip.trip.id
+                        } == true
+                    }
+                    .groupBy({ it.routeId }, { it.id })
+
+            // We don't have access to global data or representative trips here, so the only way
+            // that we can determine headsigns is by looking at the actual upcoming trips.
+            val headsignsAndPatternsToDisplayByRoute =
+                upcomingPatternsByRouteAndHeadsign
+                    .mapNotNull {
+                        val headsignsToDisplay =
+                            it.value
+                                .mapNotNull { headsignsToPatterns ->
+                                    val patternsToDisplay =
+                                        (typicalPatternsByRoute[it.key].orEmpty() +
+                                                predictedPatternsByRoute[it.key].orEmpty())
+                                            .toSet()
+                                            .intersect(headsignsToPatterns.value.toSet())
+                                    if (patternsToDisplay.isEmpty()) {
+                                        null
+                                    } else {
+                                        headsignsToPatterns.key to patternsToDisplay
+                                    }
+                                }
+                                .toMap()
+                        if (headsignsToDisplay.isEmpty()) {
+                            null
+                        } else {
+                            it.key to headsignsToDisplay
+                        }
+                    }
+                    .toMap()
+
+            // If there is only a single route with predicted or typical trips, we want to break up
+            // the grouped direction instead into individual headsign rows.
+            val firstDisplayedRoute =
+                staticData.routes.firstOrNull {
+                    it.id == headsignsAndPatternsToDisplayByRoute.keys.firstOrNull()
+                }
+            if (headsignsAndPatternsToDisplayByRoute.size == 1 && firstDisplayedRoute != null) {
+                val headsignsToDisplay =
+                    (headsignsAndPatternsToDisplayByRoute[firstDisplayedRoute.id] ?: emptyMap())
+                return headsignsToDisplay.map { (headsign, patternIds) ->
+                    val patternsForHeadsign = patternIds.mapNotNull { patternsById[it] }
+                    RealtimePatterns.ByHeadsign(
+                        NearbyStaticData.StaticPatterns.ByHeadsign(
```

Comment:
question (non-blocking): would it be possible / simplify things at all to use the `RealtimePatterns.ByHeadsign` constructor that doesn't take a `StaticPatterns.ByHeadsign`?

Reply:
I actually set it up like that initially, by adding a new constructor for `RealtimePatterns.ByHeadsign` that took a `StaticPatterns.ByDirection`, then decided that doing it like this was actually the most straightforward approach.